tokio/src/sync/broadcast.rs

```diff
@@ -883,7 +966,101 @@ impl<T: Clone> Receiver<T> {
+
+    /// Receives the next value for this receiver and the number of messages
+    /// that were sent by a sender and have not yet been received by this
+    /// receiver.
+    ///
+    /// Each [`Receiver`] handle will receive a clone of all values sent
+    /// **after** it has subscribed.
+    ///
+    /// `Err(RecvError::Closed)` is returned when all `Sender` halves have
+    /// dropped, indicating that no further values can be sent on the channel.
+    ///
+    /// If the [`Receiver`] handle falls behind, once the channel is full, newly
+    /// sent values will overwrite old values. At this point, a call to
+    /// [`recv_with_queue_status`] will return with `Err(RecvError::Lagged)` and
+    /// the [`Receiver`]'s internal cursor is updated to point to the oldest value
+    /// still held by the channel. A subsequent call to [`recv_with_queue_status`]
+    /// will return this value **unless** it has been since overwritten.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe. If `recv_with_queue_status` is used as the
+    /// event in a [`tokio::select!`](crate::select) statement and some other branch
+    /// completes first, it is guaranteed that no messages were received on this
+    /// channel.
+    ///
+    /// [`Receiver`]: crate::sync::broadcast::Receiver
+    /// [`recv_with_queue_status`]: crate::sync::broadcast::Receiver::recv_with_queue_status
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::broadcast;
+    /// use tokio::time::sleep;
+    /// use std::time::Duration;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx1) = broadcast::channel(16);
+    ///
+    ///     let handle = tokio::spawn(async move {
+    ///         sleep(Duration::from_millis(200)).await;
+    ///         assert_eq!(rx1.recv_with_queue_status().await.unwrap(), (10, 1));
+    ///         assert_eq!(rx1.recv_with_queue_status().await.unwrap(), (20, 0));
+    ///     });
+    ///
+    ///     tx.send(10).unwrap();
+    ///     tx.send(20).unwrap();
+    ///
+    ///     handle.await.unwrap();
+    /// }
+    /// ```
+    ///
+    /// Handling lag
+    ///
+    /// ```
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = broadcast::channel(2);
+    ///
+    ///     tx.send(10).unwrap();
+    ///     tx.send(20).unwrap();
+    ///     tx.send(30).unwrap();
+    ///
+    ///     // The receiver lagged behind
+    ///     assert!(rx.recv().await.is_err());
+    ///
+    ///     // At this point, we can abort or continue with lost messages
+    ///
+    ///     assert_eq!(20, rx.recv().await.unwrap());
+    ///     assert_eq!(30, rx.recv().await.unwrap());
+    /// }
+    /// ```
+    pub async fn recv_with_queue_status(&mut self) -> Result<(T, u64), RecvError> {
+        let fut = Recv::new_with_queue_status(self);
+        match fut.await {
+            Ok(rcv_output) => match rcv_output {
```

Comment:
Why does querying this information have to happen together with receiving the item? What stops us from just having a `len` method or whatever?

Reply:
I used a method because of performance costs that would result from using something like a `len` method, since we have to acquire the tail lock to get information on what the latest sent message is.